server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeQueryHelper.java

```diff
@@ -0,0 +1,257 @@
+     * Get the star-tree leaf collector
+     * This collector computes the aggregation prematurely and invokes an early termination collector
+     */
+    public static LeafBucketCollector getStarTreeLeafCollector(
+        SearchContext context,
+        ValuesSource.Numeric valuesSource,
+        LeafReaderContext ctx,
+        LeafBucketCollector sub,
+        CompositeIndexFieldInfo starTree,
+        String metric,
+        Consumer<Long> valueConsumer,
+        Runnable finalConsumer
+    ) throws IOException {
+        StarTreeValues starTreeValues = getStarTreeValues(ctx, starTree);
+        assert starTreeValues != null;
+        String fieldName = ((ValuesSource.Numeric.FieldData) valuesSource).getIndexFieldName();
+        String metricName = StarTreeUtils.fullyQualifiedFieldNameForStarTreeMetricsDocValues(starTree.getField(), fieldName, metric);
+
+        assert starTreeValues != null;
+        SortedNumericStarTreeValuesIterator valuesIterator = (SortedNumericStarTreeValuesIterator) starTreeValues.getMetricValuesIterator(
+            metricName
+        );
+        // Obtain a FixedBitSet of matched document IDs
+        FixedBitSet matchedDocIds = getStarTreeFilteredValues(context, ctx, starTreeValues);  // Assuming this method gives a FixedBitSet
+        assert matchedDocIds != null;
+
+        int numBits = matchedDocIds.length();  // Get the length of the FixedBitSet
+        if (numBits > 0) {
+            // Iterate over the FixedBitSet
+            for (int bit = matchedDocIds.nextSetBit(0); bit != -1; bit = (bit + 1 < numBits) ? matchedDocIds.nextSetBit(bit + 1) : -1) {
+                // Advance to the entryId in the valuesIterator
+                if (valuesIterator.advanceExact(bit) == false) {
+                    continue;  // Skip if no more entries
+                }
+
+                // Iterate over the values for the current entryId
+                for (int i = 0, count = valuesIterator.valuesCount(); i < count; i++) {
+                    long value = valuesIterator.nextValue();
+                    valueConsumer.accept(value); // Apply the consumer operation (e.g., max, sum)
+                }
+            }
+        }
+
+        // Call the final consumer after processing all entries
+        finalConsumer.run();
+
+        // Return a LeafBucketCollector that terminates collection
+        return new LeafBucketCollectorBase(sub, valuesSource.doubleValues(ctx)) {
+            @Override
+            public void collect(int doc, long bucket) {
+                throw new CollectionTerminatedException();
+            }
+        };
+    }
+
+    /**
+     * Get the filtered values for the star-tree query
+     * Cache the results in case of multiple aggregations (if cache is initialized)
+     * @return FixedBitSet of matched document IDs
+     */
+    public static FixedBitSet getStarTreeFilteredValues(SearchContext context, LeafReaderContext ctx, StarTreeValues starTreeValues)
+        throws IOException {
```

Comment:
This method might get called from multiple threads if we use concurrent segment search. I suspect that you would get a `ConcurrentModificationException` with enough segments.

You could avoid that by initializing an array of `FixedBitSet` (of size `context.searcher().getLeafContexts().size()` and using `ctx.ord` as the key. 

Since Lucene 10 introduces intra-segment concurrency, I think we'll eventually need a way to guarantee that only one thread per segment initializes this value.

Reply:
Refactored the value cache map to startree context. Defining it as `volatile` should ensure thread visibility and initializing it with ConcurrentHashMap to avoid `ConcurrentModificationException`. 

When upgrading to Lucene 10, I think that will warrant some work to resolve intra-segment concurrency. I think that calls out for a broader discussion in general. Most likely, only thread per [group of threads operating on same shard] will have to compute the aggregation result while the remaining will just do nothing in this case since the star-tree index presently won't be able to split as the default index in its current implementation.

Reply:
> Refactored the value cache map to startree context. Defining it as volatile should ensure thread visibility and initializing it with ConcurrentHashMap to avoid ConcurrentModificationException.

Ideally, we should be able to avoid any atomics (including `ConcurrentHashMap`). That's why I suggested preallocating an array whose size corresponds to the number of segments. If only one thread operates on each segment (which is likely also what we would want once we have intra-segment concurrency), then you can safely let each thread overwrite a `null` value in the array without worrying about `ConcurrentModificationException`, since no two threads will try touching the same value.